src/renderers/WebGLRenderer.js

```diff
@@ -1468,49 +1472,67 @@ function WebGLRenderer( parameters ) {
 
 		}
 
-		if ( material.version === materialProperties.__version ) {
+		//
 
-			if ( material.fog && materialProperties.fog !== fog ) {
+		let needsProgramChange = false;
 
-				initMaterial( material, scene, object );
+		if ( material.version === materialProperties.__version ) {
 
-			} else if ( materialProperties.environment !== environment ) {
+			if ( materialProperties.needsLights && ( materialProperties.lightsStateVersion !== lights.state.version ) ) {
 
-				initMaterial( material, scene, object );
+				needsProgramChange = true;
 
-			} else if ( materialProperties.needsLights && ( materialProperties.lightsStateVersion !== lights.state.version ) ) {
+			} else if ( materialProperties.outputEncoding !== encoding ) {
 
-				initMaterial( material, scene, object );
+				needsProgramChange = true;
 
-			} else if ( materialProperties.numClippingPlanes !== undefined &&
-				( materialProperties.numClippingPlanes !== clipping.numPlanes ||
-				materialProperties.numIntersection !== clipping.numIntersection ) ) {
+			} else if ( object.isInstancedMesh && materialProperties.instancing === false ) {
 
-				initMaterial( material, scene, object );
+				needsProgramChange = true;
 
-			} else if ( materialProperties.outputEncoding !== encoding ) {
+			} else if ( ! object.isInstancedMesh && materialProperties.instancing === true ) {
 
-				initMaterial( material, scene, object );
+				needsProgramChange = true;
 
 			} else if ( materialProperties.envMap !== envMap ) {
 
-				initMaterial( material, scene, object );
+				needsProgramChange = true;
+
+			} else if ( material.fog && materialProperties.fog !== fog ) {
+
+				needsProgramChange = true;
+
+			} else if ( materialProperties.numClippingPlanes !== undefined &&
+				( materialProperties.numClippingPlanes !== clipping.numPlanes ||
+				materialProperties.numIntersection !== clipping.numIntersection ) ) {
+
+				needsProgramChange = true;
 
 			}
 
 		} else {
 
-			initMaterial( material, scene, object );
+			needsProgramChange = true;
 			materialProperties.__version = material.version;
 
 		}
 
+		//
+
+		let program = materialProperties.currentProgram;
+
+		if ( needsProgramChange === true ) {
```

Comment:
For clarification reason, we have lately started to make such comparison more explicit. Similar to 
```js
if ( object.isMesh === true )
```
instead of 
```
if ( object.isMesh )
```

Reply:
Ok, I wasn't aware of this. A bit unusual choice, though it can make sense in the context of JavaScript 🙂